File: typescript/packages/mechanisms/multiversx/src/exact/facilitator/scheme.ts
import {
  VerifyResponse,
  PaymentRequirements,
  PaymentPayload,
  SettleResponse,
  Network,
} from '@x402/core/types'
import { SchemeNetworkFacilitator } from '@x402/core/types/mechanisms'
import { ExactMultiversXPayload } from '../../types'
import { MultiversXSigner } from '../../signer'
import { Transaction, Address, TransactionPayload } from '@multiversx/sdk-core'
import { ApiNetworkProvider } from '@multiversx/sdk-network-providers'
import {
  CHAIN_ID_DEVNET,
  CHAIN_ID_MAINNET,
  CHAIN_ID_TESTNET,
  MULTIVERSX_API_URL_DEVNET,
  MULTIVERSX_API_URL_MAINNET,
  MULTIVERSX_API_URL_TESTNET,
  MULTIVERSX_METHOD_MULTI_TRANSFER,
  MULTIVERSX_NATIVE_TOKEN,
} from '../../constants'

/**
 * MultiversX Facilitator implementation for the Exact payment scheme.
 */
export class ExactMultiversXFacilitator implements SchemeNetworkFacilitator {
  /**
   * Initializes the ExactMultiversXFacilitator.
   *
   * @param apiUrl - The MultiversX API URL to use for simulations and status checks
   * @param signer - Optional MultiversX signer for relaying transactions
   * @param signerAddress - Optional address of the signer
   */
  constructor(
    private apiUrl: string = MULTIVERSX_API_URL_DEVNET,
    private signer?: MultiversXSigner,
    private signerAddress?: string,
  ) { }

  /**
   * The scheme identifier for this facilitator.
   *
   * @returns The string 'exact'
   */
  get scheme(): string {
    return 'exact'
  }

  /**
   * The CAIP-compatible family identifier for MultiversX.
   *
   * @returns The wildcard 'multiversx:*'
   */
  get caipFamily(): string {
    return 'multiversx:*'
  }

  /**
   * Gets extra configuration for a specific network.
   *
   * @param _network - The network identifier
   * @returns An empty record as no extra config is needed by default
   */
  getExtra(_network: Network): Record<string, unknown> {
    return {}
  }

  /**
   * Gets the list of addresses authorized to sign or facilitate for a network.
   *
   * @param _network - The network identifier
   * @returns Array containing the signer address if provided
   */
  getSigners(_network: Network | string): string[] {
    if (this.signerAddress) return [this.signerAddress]
    return []
  }

  /**
   * Verifies that a payment payload matches the requirements and is cryptographically valid.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The original payment requirements to match against
   * @returns A response indicating if the payload is valid
   */
  async verify(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
  ): Promise<VerifyResponse> {
    const relayedPayload = payload.payload as unknown as ExactMultiversXPayload
    const now = Math.floor(Date.now() / 1000)

    if (
      relayedPayload.validBefore &&
      relayedPayload.validBefore > 0 &&
      now > relayedPayload.validBefore
    ) {
      return {
        isValid: false,
        invalidReason: `Payment expired (validBefore: ${relayedPayload.validBefore}, now: ${now})`,
      }
    }
    if (
      relayedPayload.validAfter &&
      relayedPayload.validAfter > 0 &&
      now < relayedPayload.validAfter
    ) {
      return {
        isValid: false,
        invalidReason: `Payment not yet valid (validAfter: ${relayedPayload.validAfter}, now: ${now})`,
      }
    }

    const expectedReceiver = requirements.payTo
    const expectedAmount = requirements.amount
    const asset = requirements.asset || MULTIVERSX_NATIVE_TOKEN

    if (asset !== MULTIVERSX_NATIVE_TOKEN) {
      if (relayedPayload.receiver !== relayedPayload.sender) {
        return {
          isValid: false,
          invalidReason: `Receiver mismatch for ESDT (expected self-send to ${relayedPayload.sender}, got ${relayedPayload.receiver})`,
        }
      }

      const parts = (relayedPayload.data || '').split('@')
      if (parts.length < 6 || parts[0] !== MULTIVERSX_METHOD_MULTI_TRANSFER) {
        return { isValid: false, invalidReason: 'Invalid ESDT transfer data format' }
      }

      const destHex = parts[1]
      const destAddr = Address.newFromBech32(expectedReceiver)
      if (destHex !== destAddr.hex()) {
        return {
          isValid: false,
          invalidReason: `Receiver mismatch in data (expected ${expectedReceiver}, got hex ${destHex})`,
        }
      }

      const tokenHex = parts[3]
      const tokenStr = Buffer.from(tokenHex, 'hex').toString('utf8')
      if (tokenStr !== asset) {
        return {
          isValid: false,
          invalidReason: `Asset mismatch (expected ${asset}, got ${tokenStr})`,
        }
      }

      const amountHex = parts[5]
      const amountBi = BigInt('0x' + amountHex)
      const expectedBi = BigInt(expectedAmount)
      if (amountBi < expectedBi) {
        return {
          isValid: false,
          invalidReason: `Amount too low (expected ${expectedAmount}, got ${amountBi.toString()})`,
        }
      }
    } else {
      if (relayedPayload.receiver !== expectedReceiver) {
        return {
          isValid: false,
          invalidReason: `Receiver mismatch (expected ${expectedReceiver}, got ${relayedPayload.receiver})`,
        }
      }

      const valBi = BigInt(relayedPayload.value)
      const expBi = BigInt(expectedAmount)
      if (valBi < expBi) {
        return {
          isValid: false,
          invalidReason: `Amount too low (expected ${expectedAmount}, got ${valBi.toString()})`,
        }
      }
    }

    const signatureValid = await this.verifySignature(relayedPayload)
    if (!signatureValid.isValid) {
      return signatureValid
    }

    const apiUrl = this.resolveApiUrl(requirements.network)
    const provider = new ApiNetworkProvider(apiUrl)
    return this.verifyViaSimulation(relayedPayload, provider)
  }

  /**
   * Broadcasts the payment transaction to the MultiversX network.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The requirements for the payment
   * @returns A response indicating if the settlement was successful and the transaction hash
   */
  async settle(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
  ): Promise<SettleResponse> {
    const relayedPayload = payload.payload as unknown as ExactMultiversXPayload
    const network = requirements.network as Network
    const payer = relayedPayload.sender
    const apiUrl = this.resolveApiUrl(network)
    const provider = new ApiNetworkProvider(apiUrl)

    try {
      const transaction = new Transaction({
        nonce: BigInt(relayedPayload.nonce),
        value: relayedPayload.value,
        receiver: Address.newFromBech32(relayedPayload.receiver),
        sender: Address.newFromBech32(relayedPayload.sender),
        gasPrice: BigInt(relayedPayload.gasPrice),
        gasLimit: relayedPayload.gasLimit,
        data: new TransactionPayload(relayedPayload.data),
        chainID: relayedPayload.chainID,
        version: relayedPayload.version,
        options: relayedPayload.options,
      })

      transaction.applySignature(Buffer.from(relayedPayload.signature || '', 'hex'))

      if (relayedPayload.relayer) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ; (transaction as any).relayer = Address.newFromBech32(relayedPayload.relayer)
        if (relayedPayload.relayerSignature) {
          transaction.applySignature(Buffer.from(relayedPayload.relayerSignature, 'hex'))
        }
      }

      const txHash = await provider.sendTransaction(transaction)

      if (!txHash) {
        return {
          success: false,
          errorReason: 'Broadcast succeeded but no hash returned',
          transaction: '',
          network,
          payer,
        }
      }

      try {
        await this.waitForTx(txHash, provider)
      } catch (waitError: unknown) {
        const err = waitError as Error
        return {
          success: false,
          errorReason: `Transaction broadcasted but wait failed: ${err.message}`,
          transaction: txHash,
          network,
          payer,
        }
      }

      return {
        success: true,
        transaction: txHash,
        network,
        payer,
      }
    } catch (e: unknown) {
      const err = e as Error
      return { success: false, errorReason: err.message, transaction: '', network, payer }
    }
  }

  /**
   * Performs an on-chain simulation of the transaction to verify its validity.
   *
   * @param payload - The payload to simulate
   * @param provider - The MultiversX API provider to use for simulations
   * @returns A verification response based on simulation results
   */
  private async verifyViaSimulation(
    payload: ExactMultiversXPayload,
    provider: ApiNetworkProvider,
  ): Promise<VerifyResponse> {
    try {
      const transaction = new Transaction({
        nonce: BigInt(payload.nonce),
        value: payload.value,
        receiver: Address.newFromBech32(payload.receiver),
        sender: Address.newFromBech32(payload.sender),
        gasLimit: payload.gasLimit,
        gasPrice: BigInt(payload.gasPrice),
        data: new TransactionPayload(payload.data),
        chainID: payload.chainID,
        version: payload.version,
        options: payload.options,
      })

      transaction.applySignature(Buffer.from(payload.signature || '', 'hex'))

      const relayerAddr = payload.relayer
      let relayerSig = payload.relayerSignature

      if (this.signer && this.signerAddress && relayerAddr === this.signerAddress && !relayerSig) {
        if (relayerAddr) {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          ; (transaction as any).relayer = Address.newFromBech32(relayerAddr)
        }
        relayerSig = await this.signer.signTransaction(transaction)
      }

      if (relayerAddr) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ; (transaction as any).relayer = Address.newFromBech32(relayerAddr)
        if (relayerSig) {
          transaction.applySignature(Buffer.from(relayerSig, 'hex'))
        }
      }

      const simResult = await provider.simulateTransaction(transaction)

      if (simResult.error) {
        return { isValid: false, invalidReason: `Simulation error: ${simResult.error}` }
      }

      const status = simResult.status
      if (status !== 'success' && status !== 'successful') {
        return {
          isValid: false,
          invalidReason: `Simulation failed: ${simResult.returnMessage || status}`,
        }
      }

      return { isValid: true }
    } catch (e: unknown) {
      const err = e as Error
      return { isValid: false, invalidReason: `Simulation exception: ${err.message}` }
    }
  }

  /**
   * Verifies the Ed25519 signature of the transaction.
   *
   * @param payload - The payload containing the signature and transaction data
   * @returns A verification response based on the cryptographic check
   */
  private async verifySignature(payload: ExactMultiversXPayload): Promise<VerifyResponse> {
    try {
      if (!payload.signature) {
        return { isValid: false, invalidReason: 'Missing signature' }
      }

      const tx = new Transaction({
        nonce: BigInt(payload.nonce),
        value: payload.value,
        receiver: Address.newFromBech32(payload.receiver),
        sender: Address.newFromBech32(payload.sender),
        gasLimit: payload.gasLimit,
        gasPrice: BigInt(payload.gasPrice),
        data: new TransactionPayload(payload.data || ''),
        chainID: payload.chainID,
        version: payload.version,
        options: payload.options,
      })

      const { TransactionComputer } = await import('@multiversx/sdk-core')
      const txComputer = new TransactionComputer()
      const serializedTx = txComputer.computeBytesForSigning(tx)

      const ed = await import('@noble/ed25519')

      const senderAddress = Address.newFromBech32(payload.sender)
      const publicKeyBytes = senderAddress.getPublicKey()
      const signatureBytes = Buffer.from(payload.signature, 'hex')

      const isValid = await ed.verifyAsync(signatureBytes, serializedTx, publicKeyBytes)

      if (!isValid) {
        return { isValid: false, invalidReason: 'Invalid Ed25519 signature' }
      }

      return { isValid: true, payer: payload.sender }
    } catch (e: unknown) {
      const err = e as Error
      return { isValid: false, invalidReason: `Signature verification failed: ${err.message}` }
    }
  }

  /**
   * Status check helper that waits until a transaction is finalized or fails.
   *
   * @param txHash - The hash of the transaction to wait for
   * @param provider - The MultiversX API provider to use for waiting
   * @throws Error if the transaction fails
   */
  private async waitForTx(txHash: string, provider: ApiNetworkProvider): Promise<void> {
    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      await (provider as any).awaitTransactionCompleted(txHash)
    } catch (e: unknown) {
      const err = e as Error
      throw new Error(`Transaction failed or wait timed out: ${err.message}`)
    }
  }

  /**
   * Resolves the MultiversX API URL for a given network identifier.
   *
   * @param network - The network identifier (e.g., 'multiversx:1', 'multiversx:D')
   * @returns The resolved API URL
   */
  private resolveApiUrl(network: string): string {
    const chainId = network.split(':')[1]
    switch (chainId) {
      case CHAIN_ID_MAINNET:
        return MULTIVERSX_API_URL_MAINNET
      case CHAIN_ID_DEVNET:
        return MULTIVERSX_API_URL_DEVNET
      case CHAIN_ID_TESTNET:
        return MULTIVERSX_API_URL_TESTNET
      default:
        return this.apiUrl
    }
  }
}

File: go/mechanisms/multiversx/exact/facilitator/scheme.go
package facilitator

import (
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"math/big"
	"net/http"
	"strings"
	"time"

	"github.com/multiversx/mx-chain-core-go/data/transaction"
	"github.com/multiversx/mx-sdk-go/blockchain"
	"github.com/multiversx/mx-sdk-go/core"
	"github.com/multiversx/mx-sdk-go/data"

	"github.com/coinbase/x402/go/mechanisms/multiversx"

	x402 "github.com/coinbase/x402/go"
	"github.com/coinbase/x402/go/types"
)

// Proxy defines the interface for interacting with MultiversX blockchain
type Proxy interface {
	GetTransactionStatus(ctx context.Context, hash string) (string, error)
	GetTransactionInfo(ctx context.Context, hash string) (*data.TransactionInfo, error)
	GetTransactionInfoWithResults(ctx context.Context, hash string) (*data.TransactionInfo, error)
	GetAccount(ctx context.Context, address core.AddressHandler) (*data.Account, error)
	SendTransaction(ctx context.Context, tx *transaction.FrontendTransaction) (string, error)
}

// ExactMultiversXScheme implements SchemeNetworkFacilitator
type ExactMultiversXScheme struct {
	config multiversx.NetworkConfig
	proxy  Proxy
	signer multiversx.FacilitatorMultiversXSigner
}

func NewExactMultiversXScheme(apiUrl string, signer multiversx.FacilitatorMultiversXSigner) (*ExactMultiversXScheme, error) {
	args := blockchain.ArgsProxy{
		ProxyURL:            apiUrl,
		Client:              nil,
		SameScState:         false,
		ShouldBeSynced:      false,
		FinalityCheck:       false,
		EntityType:          core.Proxy,
		CacheExpirationTime: time.Minute,
	}
	proxy, err := blockchain.NewProxy(args)
	if err != nil {
		return nil, fmt.Errorf("failed to create proxy: %w", err)
	}

	p, ok := interface{}(proxy).(Proxy)
	if !ok {
		return nil, fmt.Errorf("proxy does not implement the required interface")
	}

	return &ExactMultiversXScheme{
		config: multiversx.NetworkConfig{ApiUrl: apiUrl},
		proxy:  p,
		signer: signer,
	}, nil
}

func (s *ExactMultiversXScheme) Scheme() string {
	return multiversx.SchemeExact
}

func (s *ExactMultiversXScheme) CaipFamily() string {
	return "multiversx:*"
}

func (s *ExactMultiversXScheme) GetExtra(network x402.Network) map[string]interface{} {
	return nil
}

func (s *ExactMultiversXScheme) GetSigners(network x402.Network) []string {
	if s.signer != nil {
		return s.signer.GetAddresses()
	}
	return []string{}
}

func (s *ExactMultiversXScheme) Verify(ctx context.Context, payload types.PaymentPayload, requirements types.PaymentRequirements) (*x402.VerifyResponse, error) {
	relayedPayloadPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewVerifyError(x402.ErrCodeInvalidPayment, "", "multiversx", fmt.Errorf("invalid payload format: %v", err))
	}
	relayedPayload := *relayedPayloadPtr

	isValid, err := multiversx.VerifyPayment(ctx, relayedPayload, requirements, s.verifyViaSimulation)
	if err != nil {
		return nil, err
	}
	if !isValid {
		return nil, x402.NewVerifyError(x402.ErrCodeSignatureInvalid, relayedPayload.Sender, "multiversx", nil)
	}

	now := uint64(time.Now().Unix())
	if relayedPayload.ValidBefore > 0 && now > relayedPayload.ValidBefore {
		return nil, fmt.Errorf("payment expired (validBefore: %d, now: %d)", relayedPayload.ValidBefore, now)
	}
	if relayedPayload.ValidAfter > 0 && now < relayedPayload.ValidAfter {
		return nil, fmt.Errorf("payment not yet valid (validAfter: %d, now: %d)", relayedPayload.ValidAfter, now)
	}

	expectedReceiver := requirements.PayTo
	expectedAmount := requirements.Amount
	if expectedAmount == "" {
		return nil, errors.New("requirement amount is empty")
	}

	reqAsset := requirements.Asset
	if reqAsset == "" {
		return nil, errors.New("requirement asset is required")
	}

	txData := relayedPayload
	transferMethod, _ := requirements.Extra["assetTransferMethod"].(string)

	if reqAsset == multiversx.NativeTokenTicker && transferMethod != multiversx.TransferMethodESDT {
		if txData.Receiver != expectedReceiver {
			return nil, fmt.Errorf("receiver mismatch: expected %s, got %s", expectedReceiver, txData.Receiver)
		}
		if !multiversx.CheckBigInt(txData.Value, expectedAmount) {
			return nil, fmt.Errorf("amount mismatch: expected %s, got %s", expectedAmount, txData.Value)
		}
	} else {
		parts := strings.Split(txData.Data, "@")
		if len(parts) < 6 || parts[0] != "MultiESDTNFTTransfer" {
			return nil, errors.New("invalid ESDT transfer data format (expected MultiESDTNFTTransfer)")
		}

		destHex := parts[1]
		if !multiversx.IsValidHex(destHex) {
			return nil, fmt.Errorf("invalid receiver hex")
		}

		expectedAddr, err := data.NewAddressFromBech32String(expectedReceiver)
		if err != nil {
			return nil, fmt.Errorf("invalid expected receiver format: %v", err)
		}
		expectedHex := hex.EncodeToString(expectedAddr.AddressBytes())

		if destHex != expectedHex {
			return nil, fmt.Errorf("receiver mismatch: encoded destination %s does not match requirement %s", destHex, expectedReceiver)
		}

		tokenBytes, err := hex.DecodeString(parts[3])
		if err != nil {
			return nil, fmt.Errorf("invalid token hex")
		}
		if string(tokenBytes) != reqAsset {
			return nil, fmt.Errorf("asset mismatch: expected %s, got %s", reqAsset, string(tokenBytes))
		}

		amountBytes, err := hex.DecodeString(parts[5])
		if err != nil {
			return nil, fmt.Errorf("invalid amount hex")
		}
		amountBig := new(big.Int).SetBytes(amountBytes)
		expectedBig, ok := new(big.Int).SetString(expectedAmount, 10)
		if !ok {
			return nil, fmt.Errorf("invalid expected amount: %s", expectedAmount)
		}
		if amountBig.Cmp(expectedBig) < 0 {
			return nil, fmt.Errorf("amount too low or invalid")
		}
	}

	return &x402.VerifyResponse{
		IsValid: true,
	}, nil
}

func (s *ExactMultiversXScheme) Settle(ctx context.Context, payload types.PaymentPayload, requirements types.PaymentRequirements) (*x402.SettleResponse, error) {
	relayedPayloadPtr, err := multiversx.PayloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewSettleError("invalid_payload", "", "multiversx", "", err)
	}
	relayedPayload := *relayedPayloadPtr

	tx := relayedPayload.ToTransaction()

	var hash string
	if s.signer != nil {
		// Native Relayed V3 (MIP-12)
		// 1. Fetch facilitator (relayer) addresses
		addresses := s.signer.GetAddresses()
		if len(addresses) == 0 {
			return nil, x402.NewSettleError("no_signer_address", relayedPayload.Sender, "multiversx", "", errors.New("signer has no addresses"))
		}
		facilitatorAddr := addresses[0]

		// 2. Prepare Relayed V3 transaction
		// The inner transaction is what the client signed. We just add relayer info.
		// We MUST NOT change the original sender, receiver, value, or data.
		// We also must not change the GasLimit if it already includes the relay fee.
		tx.RelayerAddr = facilitatorAddr
		tx.Version = 2 // Relayed V3 uses version 2

		// 3. Faciliator (Relayer) signs the transaction
		sig, err := s.signer.Sign(ctx, &tx)
		if err != nil {
			return nil, x402.NewSettleError("signing_failed", relayedPayload.Sender, "multiversx", "", err)
		}
		tx.RelayerSignature = sig

		// 4. Send the Native Relayed V3 transaction
		hash, err = s.signer.SendTransaction(ctx, &tx)
		log.Printf("[MVX] RelayedV3 Tx Sent. Hash: %s", hash)
	} else {
		hash, err = s.proxy.SendTransaction(ctx, &tx)
		log.Printf("[MVX] Direct Tx Sent. Hash: %s", hash)
	}

	if err != nil {
		return nil, x402.NewSettleError("broadcast_failed", relayedPayload.Sender, "multiversx", "", err)
	}

	log.Printf("[MVX] Waiting for tx: %s", hash)
	if err := s.waitForTx(ctx, hash); err != nil {
		return nil, x402.NewSettleError("tx_failed", relayedPayload.Sender, "multiversx", hash, err)
	}

	return &x402.SettleResponse{
		Success:     true,
		Transaction: hash,
	}, nil
}

// waitForTx polls the transaction status using the proxy
func (s *ExactMultiversXScheme) waitForTx(ctx context.Context, txHash string) error {
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()

	// Wait up to 120 seconds
	timeout := time.After(120 * time.Second)

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-timeout:
			return fmt.Errorf("timeout waiting for tx %s", txHash)
		case <-ticker.C:
			status, err := s.getTransactionStatus(ctx, txHash)
			if err != nil {
				continue // retry on transient errors
			}

			switch status {
			case "success", "successful", "executed":
				return nil
			case "fail", "failed", "invalid":
				return fmt.Errorf("transaction failed with status: %s", status)
			case "pending", "processing", "received", "partially-executed":
				continue
			default:
				// t.Logf("Unknown transaction status: %s", status)
				continue
			}
		}
	}
}

// getTransactionStatus fetches status via the proxy engine
func (s *ExactMultiversXScheme) getTransactionStatus(ctx context.Context, txHash string) (string, error) {
	status, err := s.proxy.GetTransactionStatus(ctx, txHash)
	if err != nil {
		return "", err
	}

	if status == "fail" || status == "failed" || status == "invalid" {
		txInfo, err := s.proxy.GetTransactionInfo(ctx, txHash)
		if err == nil && txInfo.Error != "" {
			log.Printf("[MVX] Tx %s Status: %s (error: %s)", txHash, status, txInfo.Error)
			return fmt.Sprintf("%s (error: %s)", status, txInfo.Error), nil
		}
	}

	log.Printf("[MVX] Tx %s Status: %s", txHash, status)
	return status, nil
}

func (s *ExactMultiversXScheme) verifyViaSimulation(payload multiversx.ExactRelayedPayload) (string, error) {
	tx := payload.ToTransaction()
	if tx.Version >= 2 && tx.RelayerAddr != "" && s.signer != nil {
		// Attempt to sign as relayer if we hold the key
		addresses := s.signer.GetAddresses()
		for _, addr := range addresses {
			if addr == tx.RelayerAddr {
				// We are the relayer
				sig, err := s.signer.Sign(context.TODO(), &tx)
				if err != nil {
					return "", fmt.Errorf("failed to sign as relayer: %w", err)
				}
				if tx.RelayerSignature == "" {
					tx.RelayerSignature = sig
				}
				break
			}
		}

		// Ensure empty string if still missing (simulation requirement if not signed)
		if tx.RelayerSignature == "" {
			tx.RelayerSignature = ""
		}
	}

	txBytes, err := json.Marshal(tx)
	if err != nil {
		return "", err
	}
	fmt.Printf("[MVX] Simulation JSON: %s\n", string(txBytes))

	url := fmt.Sprintf("%s/transaction/simulate", s.config.ApiUrl)

	req, err := http.NewRequestWithContext(context.Background(), "POST", url, bytes.NewBuffer(txBytes))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("simulation api error: %s - %s", resp.Status, string(body))
	}

	var res struct {
		Data struct {
			Result struct {
				Status string `json:"status"`
				Hash   string `json:"hash"`
			} `json:"result"`
		} `json:"data"`
		Error string `json:"error"`
		Code  string `json:"code"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return "", err
	}
	if res.Error != "" {
		return "", errors.New(res.Error)
	}

	if res.Code == "successful" {
		hash := res.Data.Result.Hash
		if hash == "" {
			hash = "simulated"
		}
		return hash, nil
	}

	if res.Data.Result.Status != "success" && res.Data.Result.Status != "successful" {
		return "", fmt.Errorf("simulation status not success: %s (code: %s)", res.Data.Result.Status, res.Code)
	}

	return res.Data.Result.Hash, nil
}
